service/storage_service.cc

```diff
@@ -20,6 +20,10 @@
 #include "db/system_distributed_keyspace.hh"
 #include "db/consistency_level.hh"
 #include <seastar/core/smp.hh>
+#include "mutation/canonical_mutation.hh"
```

Comment:
> In a multi-DC cluster with 100ms latencies between DCs, which means ~200ms for
each command, because we first perform a barrier in start_operation and then commit
the command. Thus this operation will take about 1.6s which is incomparably shorter
than bootstrapping operation (bootstrapping is quick if there is no data in the cluster,
but usually if one has 100 nodes they have tons of data, so indeed streaming/repair
will take much longer (hours/days)).

With tablets the bootstrap time will not depend on the amount of data. A node will be added without streaming and data will be migrated to it later by a load balancer.

> The alternative solution to sequential sending could be sending these commands
concurrently. But it would be complex, it's not clear how to do it with `group0_guard`
and this operation's time is not a problem.

Now the data is send unguarded, so is there still a problem?

I would even do writes through storage proxy with CL=ALL (retrying each failed write individually). Well this should not be a regular CL=ALL (since we may want to ignore some nodes), but something that gets a list of nodes and use them as CL=ALL targets.

Reply:
> Each mutation is sent in separate raft command and is retried as long as `group0_concurrent_modification` occurs.

This is no longer the case since the exception is not thrown any longer, no?

Reply:
> > Now the data is send unguarded, so is there still a problem?
> 
> We shouldn't send all mutations at once, we should limit concurrency based on the number of nodes, size and number of mutations. That's still a bit of complexity.

We have `max_concurrent_for_each`, but I do not understand why should we limit concurrency here especially dynamically.

> 
> I would prefer we do it the simpler way now (sequentially). It can be optimized in the future if we really think that the <=2 seconds is so problematic. But it happens only in very large clusters anyway.
> 
> > I would even do writes through storage proxy with CL=ALL (retrying each failed write individually). Well this should not be a regular CL=ALL (since we may want to ignore some nodes), but something that gets a list of nodes and use them as CL=ALL targets.
> 
> Every member of group 0 must observe the data inside CDC_GENERATIONS_V3 when it gets the `topology_change` that commits the generation. So you need to ensure that your CL=ALL targets include exactly every member of group 0. Furthermore, that data needs to be streamed to new nodes when they join group 0, and if we didn't make it part of group 0 as you suggest (IIUC) then we would have to stream it **before** they join group 0 (because as soon as they join, their topology state points to the data in the table).

If you do it with CL=ALL and retry each failure it is guarantied that all nodes will have it when `topology_change`  runs. I am not sure streaming is needed. We do not (will not) add nodes while we add nodes. Wrote it in another comment.

> 
> So why not do it like in this PR, and keep the data in group 0, which solves all these problems automagically? If you suggest an alternative you should explain why, and what are the downsides of current solution that your alternative solves.

This `write_mutation` command looks like a one use hack. If the goal is serialization there are other ways.

Reply:
> > We have max_concurrent_for_each, but I do not understand why should we limit concurrency here especially dynamically.
> 
> Limiting memory consumption - e.g. suppose you have 100 nodes and 30MB generation, if you send the write concurrently to everyone you could have a memory usage spike up to 3GB IIUC.

This is the job of raft to limit the consumption given that we use raft (and raft does limit log length). If we move to storage proxy a static concurrency is sufficient. Sending one mutation at a time is static concurrency of one anyway. 

> 
> > If you do it with CL=ALL and retry each failure it is guarantied that all nodes will have it when topology_change runs. I am not sure streaming is needed. We do not (will not) add nodes while we add nodes.
> 
> New nodes will join group 0 later. They will need to access this data too. So let's distribute it through group 0. In general if something in group 0 relies on some data it's simplest to put that data in group 0.

They do not need partial data, and last generation we already transfer as part of the snapshot.

> 
> Reading the data even by the current members is simplified when it's written using group 0, because then it is living in a local table and to obtain it, we just do a local read. We don't need to do CL=ONE distributed reads like we currently do with CDC_GENERATIONS_V2, which requires retry loops and so on.

I propose to write it into the new table, just use storage proxy code for that. I clarified that when I say CL=ALL I do not mean real CL=ALL:

"Well this should not be a regular CL=ALL (since we may want to ignore some nodes), but something that gets a list of nodes and use them as CL=ALL targets."

I propose to write the code similar to what we have for batches in storage_proxy now. It will construct specially crafted mutation_write_handler with all the targets that we want to write into.

Reply:
> They do not need partial data, and last generation we already transfer as part of the snapshot.

This data becomes the "last generation" once we send all of it.

Reply:
> I propose to write the code similar to what we have for batches in storage_proxy now. It will construct specially crafted mutation_write_handler with all the targets that we want to write into.

Ah, and implementing special storage_proxy logic for being able to do distributed writes to local tables with a specially crafted `mutation_write_handler` with selected targets is simpler than adding a `write_mutations` command. I see.

Reply:
> > I propose to write the code similar to what we have for batches in storage_proxy now. It will construct specially crafted mutation_write_handler with all the targets that we want to write into.
> 
> Ah, and implementing special storage_proxy logic for being able to do distributed writes to local tables with a specially crafted `mutation_write_handler` with selected targets is simpler than adding a `write_mutations` command. I see.

It is. The code will be short (handlers are extensible to support such use case). It will be more efficient. And it will not require all the dense here with ` in_progress_cdc_generation_id` here. Why do you directly assume it is more complicated?

Reply:
First of all, "complicated" doesn't mean that it requires boilerplate. "Complicated" means you have to think harder on why a particular solution is correct - at least that's the most important consideration IMO.

So if we have two writes (one for writing data, one for commiting CDC generation) which need to go to all group 0 members, you do the second through group 0 command, and the second depends on the first, how do we ensure that the second observes the first? The obvious answer is to do the first write through group 0 too.

The nonobvious answer is to do the first write using a separate mechanism, then rely on a fact which will be hopefully documented in a comment somewhere, that we do not join new members to group 0 while we're creating a new CDC generation, and then pray that this assumption never changes in the future. Actually even now I'm not sure if this assumption is true, it probably isn't to be honest, because we can call the "repair CDC generation API" and concurrently start joining a new node and this new group 0 member needs to get the CDC genetration created by the API.

So at least considering which solution is easier to understand and prove correct - IMO my solution is simpler.

Regarding boilerplate, there is a very small amount of it in this solution, as you can see in the PR - add a new command type (which we may find useful in the future too, e.g. for broadcast tables) and a new field to topology state.

In your solution, I am not able to estimate how much boilerplate will be needed. And it's also not simple boilerplate, because we invent new things like being able to do distributed writes to a local replication strategy table, which is pretty confusing.

Reply:
So IIUC the correctness argument you provide is -- we do global RPCs from the coordinator already, and this must be correct or our design is broken, so let's also introduce a way to do a global write through `storage_proxy` and use it, it will be correct by extension.

My correctness argument is - both writes are done through group 0 so the second must observe the first.

Maybe we need an external arbiter to decide which argument is simpler.

And you didn't address my amount-of-boilerplate concern. It's hard for me to believe that extending storage_proxy with this new mechanism will require less boilerplate than using `write_mutations`.

Reply:
> So IIUC the correctness argument you provide is -- we do global RPCs from the coordinator already, and this must be correct or our design is broken, so let's also introduce a way to do a global write through `storage_proxy` and use it, it will be correct by extension.

Global write is not different then global RPC. Fencing rpc also does writes. My correctness argument is that we design the system to be correct this way and any additional mechanism is unnecessary complication. Stick to the design.

> 
> My correctness argument is - both writes are done through group 0 so the second must observe the first.

Why not do everything through grioup0 them. We _can_ do RPCs through it. It is just unnecessary complication. We do not need order between CDC table writes, so why do you insist writing them through raft just in case?

> 
> Maybe we need an external arbiter to decide which argument is simpler.
> 
> And you didn't address my amount-of-boilerplate concern. It's hard for me to believe that extending storage_proxy with this new mechanism will require less boilerplate than using `write_mutations`.

Because we will not know until we try, but we have a lot of specialized write handlers (batch, views, hints, paxos), so the code is designed to be extendible (albeit  through an awkward templated methods).

Reply:
> Global write is not different then global RPC. 

In general it is different. A global write must propagate during topology changes too. Global RPC is only done on the current set of members.

In this case you made the point that we don't need to propagate this particular global write during topology changes, so we can reduce it to a global RPC.

> Why not do everything through grioup0 them. We can do RPCs through it. It is just unnecessary complication. 

I would actually prefer doing it that way if we could, but we can't due to one reason - snapshots. Because of snapshots there is no guarantee that every node will call `apply()`.

This is another difference between a global write and a global RPC. Global write can be propagated through a global RPC for some nodes and by a snapshot for others.

But this is off-topic.

> We do not need order between CDC table writes, so why do you insist writing them through raft just in case?

We **do** need order but not between the CDC table writes within each other, we need order between the set of CDC table writes as a whole and the last `topology_change` which updates `topology`, pointing to the data that was written by the CDC table writes. And because of that order requirement, I claim using group 0 for those writes is just the most obvious thing.

Reply:
> > Global write is not different then global RPC.
> 
> In general it is different. A global write must propagate during topology changes too. Global RPC is only done on the current set of members.
> 
> In this case you made the point that we don't need to propagate this particular global write during topology changes, so we can reduce it to a global RPC.

Exactly. IMO this is what allows for simplification.

> 
> > Why not do everything through grioup0 them. We can do RPCs through it. It is just unnecessary complication.
> 
> I would actually prefer doing it that way if we could, but we can't due to one reason - snapshots. Because of snapshots there is no guarantee that every node will call `apply()`.
> 
> This is another difference between a global write and a global RPC. Global write can be propagated through a global RPC for some nodes and by a snapshot for others.
> 
> But this is off-topic.

It is. Otherwise we could have argue how RPCs can be encoded in a snapshot as well :)

> 
> > We do not need order between CDC table writes, so why do you insist writing them through raft just in case?
> 
> we need order between the set of CDC table writes as a whole and the last `topology_change`

Exactly my point. So why not do those set of write in most efficient way (since no order is needed) and make sure that there is an order between side effect of those writes and last `topology_change` command like we do for any other RPC with side effects?

Reply:
>  So why not do those set of write in most efficient way (since no order is needed) 

If that's the problem, we can do it through group 0 too, since they are unguarded, we can send these commands concurrently.

BTW. the global RPCs we do from the coordinator currently are only contacting NORMAL nodes, which is probably a bug - we should be contacting bootstrapping node(s) as well. Definitely if we do the CDC generation write through a global RPC we need to do it on bootstrapping node too.

Ayyy More complications. Why not just do the write through group 0 like it's supposed to be used...

Reply:
> > So why not do those set of write in most efficient way (since no order is needed)
> 
> If that's the problem, we can do it through group 0 too, since they are unguarded, we can send these commands concurrently.
> 

That what I asked in the review. But storage proxy will still be more efficient in multi-dc setup.